src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs

```diff
@@ -15931,22 +16030,346 @@ public static Vector512<T> Invoke(Vector512<T> x)
+            private const double DOUBLE_C3 = -1.984126984120184E-4;
+            private const double DOUBLE_C4 = 2.7557319210152756E-6;
+            private const double DOUBLE_C5 = -2.5052106798274616E-8;
+            private const double DOUBLE_C6 = 1.6058936490373254E-10;
+            private const double DOUBLE_C7 = -7.642917806937501E-13;
+            private const double DOUBLE_C8 = 2.7204790963151784E-15;
+            internal const ulong DOUBLE_MAX_ARG = 0x4160000000000000ul; // 0x1p23
+
+            public static bool Vectorizable => typeof(T) == typeof(float) || typeof(T) == typeof(double);
+
             public static T Invoke(T x) => T.Cos(x);
-            public static Vector128<T> Invoke(Vector128<T> x) => throw new NotSupportedException();
-            public static Vector256<T> Invoke(Vector256<T> x) => throw new NotSupportedException();
-            public static Vector512<T> Invoke(Vector512<T> x) => throw new NotSupportedException();
+
+            public static Vector128<T> Invoke(Vector128<T> x)
+            {
+                if (typeof(T) == typeof(float))
+                {
+                    Vector128<uint> uxMasked = x.AsUInt32() & Vector128.Create(SINGLE_SIGN_MASK);
+
+                    // Remove sign from input
+                    Vector128<float> dinput = uxMasked.AsSingle();
+
+                    // n = int ((x + pi/2) /pi) - 0.5
+                    Vector128<float> almHuge = Vector128.Create(SINGLE_ALM_HUGE);
+                    Vector128<float> dn = ((dinput + Vector128.Create(float.Pi / 2)) * Vector128.Create(1 / float.Pi)) + almHuge;
+                    Vector128<uint> n = dn.AsUInt32();
+                    dn = dn - almHuge - Vector128.Create(0.5f);
+
+                    // frac = x - (n*pi)
+                    Vector128<float> frac = dinput + (dn * Vector128.Create(-float.Pi)) + (dn * Vector128.Create(SINGLE_PI_TAIL1)) + (dn * Vector128.Create(SINGLE_PI_TAIL2));
+
+                    // Check if n is odd or not
+                    // Compute sin(f) using the polynomial: x*(1+C1*x^2+C2*x^4+C3*x^6+C4*x^8)
+                    // If n is odd, result is negative
+                    Vector128<uint> odd = n << 31;
+                    Vector128<float> poly = PolyEvalOdd9(frac, Vector128<float>.One, Vector128.Create(SINGLE_C1), Vector128.Create(SINGLE_C2), Vector128.Create(SINGLE_C3), Vector128.Create(SINGLE_C4));
+                    Vector128<float> result = (poly.AsUInt32() ^ odd).AsSingle();
+
+                    // If any of the input values are greater than ARG_MAX, call scalar
+                    if (Vector128.GreaterThanAny(uxMasked, Vector128.Create(SINGLE_MAX_ARG)))
+                    {
+                        Vector128<float> floats = x.AsSingle();
+                        result = Vector128.Create(float.Cos(floats[0]), float.Cos(floats[1]), float.Cos(floats[2]), float.Cos(floats[3]));
+                    }
+
+                    return result.As<float, T>();
```

Comment:
nit: We should probably follow the same logic as used for `ExpOperator` (and ideally expose the implementation for `Vector64/128/256/512` simultaneously, since that's approved surface area).

This is a particular case where having `Invoke` do simple dispatch to a helper function that handles `float` or that handles `double` is "goodness", because the JIT is less likely to run into inlining limitations from seeing a "huge" method containing tons of logic and many branches.

Reply:
> and ideally expose the implementation for Vector64/128/256/512 simultaneously, since that's approved surface area

I'm not prioritizing that. Those are nice-to-have from my perspective.

> We should probably follow the same logic as used for ExpOperator

Can you elaborate?

Reply:
> I'm not prioritizing that. Those are nice-to-have from my perspective.

I was commenting on it because it's basically the same amount of work. We've already implemented the algorithm here, so to finish out the surface area on .NET 9, it's basically just copy/pasting the code you've already written into the `VectorMath` helper class.

I won't push if you really don't want to do it here though.

> Can you elaborate?

Instead of:
```csharp
public static Vector128<T> Invoke(Vector128<T> x)
{
    if (typeof(T) == typeof(double))
    {
        // lot of code, with many locals, branches, etc
    }
    else
    {
        Debug.Assert(typeof(T) == typeof(float));
        // lot of code, with many locals, branches, etc
    }
}
```

do it as:
```csharp
public static Vector128<T> Invoke(Vector128<T> x)
{
    if (typeof(T) == typeof(double))
    {
        return ExpOperatorDouble.Invoke(x.AsDouble()).As<double, T>();
    }
    else
    {
        Debug.Assert(typeof(T) == typeof(float));
        return ExpOperatorSingle.Invoke(x.AsSingle()).As<float, T>();
    }
}
```

We have known limitations around inlining large and overly complex methods on all runtimes, but this is particularly problematic for Mono and the interpreter.

By breaking it apart like this, we make it easier for all runtimes to see that `Invoke` is a trivial method (only containing typeof checks for branches with a single non-inspectable method call and 2 intrinsic calls per path). It allows the JIT to then make an easy decision to inline `Invoke` and do early dead code elimination without having to worry about complex flow analysis or dealing with many additional locals.

This can often improve the overall inlining and JIT throughput, leading to better overall codegen while also setting us up in the general pattern we want to eventually defer to `Vector128.Cos` for .NET 9 (once those APIs are also implemented), allowing fewer diffs later on.

Reply:
> I was commenting on it because it's basically the same amount of work. 

It's not, though. More surface area, more tests, more involved components, more moving pieces. I'm not going to do it here.

Reply:
I split Sin/CosOperator into float/double flavors.